src/basic/virt.c

```diff
@@ -235,8 +235,28 @@ static int detect_vm_dmi(void) {
 
         /* The DMI vendor tables in /sys/class/dmi/id don't help us distinguish between Amazon EC2
          * virtual machines and bare-metal instances, so we need to look at SMBIOS. */
-        if (r == VIRTUALIZATION_AMAZON && detect_vm_smbios() == SMBIOS_VM_BIT_UNSET)
-                return VIRTUALIZATION_NONE;
+        if (r == VIRTUALIZATION_AMAZON) {
+                switch (detect_vm_smbios()) {
+                case SMBIOS_VM_BIT_UNKNOWN: {
+                        /* The DMI information we are after is only accessible to the root user,
+                         * so we fallback to using the product name which is less restricted
+                         * to distinguish metal systems from virtualized instances */
+                        _cleanup_free_ char *s = NULL;
+                        int k = read_full_virtual_file("/sys/class/dmi/id/product_name", &s, NULL);
```

Comment:
Please add one blank line between the variable declaration and the others. Also, it is not necessary it introduce `k`:
```
if (r == VIRTUALIZATION_AMAZON)
    switch(detect_vm_bios) {
    case SMBIOS_VM_BIT_SET:
        return VIRTUALIZATION_AMAZON;
    case SMBIOS_VM_BIT_UNSET:
        return VIRTUALIZATION_NONE;
    case SMBIOS_VM_BIT_UNKNOWN: {
        _cleanup_free_ char *s = NULL;

        r = read_full_virtual_file(...);
        if (r < 0) {
            log_debug_errno(r, "..., assuming virtualized: %m");
            return VIRTUALIZATION_AMAZON;
        }
        if (endswith(s, ".metal\n"))
            return VIRTUALIZATION_NONE;
        else
            return VIRTUALIZATION_AMAZON;
    }
    default:
        assert_not_reached();
    }
```

Reply:
Ok.

Reply:
Please return earlier as I wrote in the above. It is not necessary to continue the function after this `switch()`.

Reply:
Gosh this is some serious case of bike shed painting... I'll do it though